### ISSUE TITLE
Quad Rendering

### ISSUE THREAD

**Issue by ADormant** [tapcio/PeteOpenGL2Tweak#5]
Is this doable through PeteTweak interface?
Though this particular implementation isn't 100% correct yet.
https://github.com/simias/rustation/commit/4b0b5f28b01c3cc56c8f5789b06fcd276153dbc5

**Comment by Papermanzero** [comment]
This would be a great addition!

**Comment by Nucleoprotein** [comment]
"Quad Rendering" will do nothing really, i think you think "Quad Rendering" is perspective correction: https://en.wikipedia.org/wiki/Texture_mapping#Perspective_correctness
For perspective correction plugin needs Z, no Z, no perspective correction. So GTE Acc hack need to be fixed before this.

**Comment by Papermanzero** [comment]
Yes, that's right. Let's call it perspective correction then. :-)

**Comment by ADormant** [comment]
@tapcio Actually quad rendering and perspective-correct texture mapping are separate features in gpubladesoft and can be used at the same time(but mapping isn't fully implemented yet).  Quads give perspective correction a la Saturn which didn't have perspective correct mapping.

**Comment by Nucleoprotein** [comment]
I do not know how quads change anything in rendering because is only marge two tingles to quad so texture will look same, ie. perspective will be wrong if Z is not used to correct it.

**Comment by ADormant** [comment]
Apparently it does somehow for both Saturn and PS1. Maybe @simias can explain how quad based perspective correction works.  Of course quads+Z is even better.
![need for speed quad](https://cloud.githubusercontent.com/assets/10523326/13847312/ea30036c-ec4c-11e5-90de-52dcadb09162.gif)
![test drive quad](https://cloud.githubusercontent.com/assets/10523326/13847412/6af428b6-ec4d-11e5-9818-aa79b858b32b.gif)

**Comment by ADormant** [comment]
Seems it's because quads are more resistant to warping.
http://www.reedbeta.com/blog/2012/05/26/quadrilateral-interpolation-part-1/

**Comment by simias** [comment]
If you map textures to quads instead of triangles you can do bilinear mapping which is still not perspective correct but generally looks less warped than affine mapping on triangles, especially for things like walls and floors.

Affine mapping on triangle-mapped quads creates some very obvious distortion along the diagonal which creates these very weird looking seams and makes straight lines bend halfway across the quad. Bilinear mapping on a quad wouldn't have this artifact. It'll still "stretch" the texture because of the lack of perspective correction but it won't bend the texture.

In rustation's branch @ADormant mentioned I implemented a different approach where I attempted to guess the perspective of the quad assuming that it's a rectangle in 3D space. It works well for walls and floors but gives weird results when the quad is not meant to be rectangular and the algorithm ends up guessing a wrong perspective.
